common/src/main/java/org/conscrypt/HpkeParameterSpec.java

```diff
@@ -0,0 +1,655 @@
+        this.authKey = builder.authKey;
+        this.l = builder.l;
+        this.encrypting = builder.encrypting;
+        this.exporting = builder.exporting;
+        this.mode = builder.mode;
+    }
+
+    /**
+     * Returns the HPKE algorithm identifier.
+     *
+     * @return algorithm identifier specifying the KEM, KDF, and AEAD configured.
+     */
+    public HpkeAlgorithmIdentifier getAlgorithmIdentifier() {
+        return algorithmIdentifier;
+    }
+
+    /**
+     * Returns the encapsulated key.
+     *
+     * @return enc creates a new array each time this method is called.
+     */
+    public byte[] getEnc() {
+        return enc == null ? null : enc.clone();
+    }
+
+    /**
+     * Returns the export length (L).
+     *
+     * @return L.
+     */
+    public int getL() {
+        return l;
+    }
+
+    /**
+     * Returns the caller's info.
+     *
+     * @return info creates a new array each time this method is called.
+     */
+    public byte[] getInfo() {
+        return info == null ? null : info.clone();
+    }
+
+    /**
+     * Returns the Initialization Vector (IV). The IV is a sender key value to be configured for
+     * testing purposes only. Setting an IV avoids output randomness making it easier to validate
+     * encryption results.
+     *
+     * @return IV creates a new array each time this method is called.
+     */
+    public byte[] getIv() {
+        return iv == null ? null : iv.clone();
+    }
+
+    /**
+     * Returns the pre-shared key (PSK) held by both the sender and recipient.
+     *
+     * @return psk creates a new array each time this method is called.
+     */
+    public byte[] getPsk() {
+        return psk == null ? null : psk.clone();
+    }
+
+    /**
+     * Returns an identifier for the PSK.
+     *
+     * @return psk identifier.
+     */
+    public byte[] getPskId() {
+        return pskId == null ? null : pskId.clone();
+    }
+
+    /**
+     * Returns key used for authentication using an asymmetric key, the key could be a private or
+     * public key, depending on the mode. If encrypting or send export, then the key should be a
+     * secret key. If decrypting or receive export, then the key should be a public key.
+     *
+     * @return public/private key for authentication using an asymmetric key.
+     */
+    public byte[] getAuthKey() {
+        return authKey == null ? null : authKey.clone();
+    }
+
+    /**
+     * Returns the HPKE mode configured. The possible modes are BASE, PSK, AUTH, and AUTH_PSK.
+     *
+     * @return mode
+     */
+    public Mode getMode() {
+        return mode;
+    }
+
+    /**
+     * Depending on how the {@link Builder} is initialized, returns a boolean telling if the API to
+     * be used is either Secret Export or an Encryption/Decryption API. If true, then the API is
+     * Secret Export.
+     *
+     * @return API configured, either Secret Export or Encryption/Decryption.
+     */
+    public boolean isExporting() {
```

Comment:
What's this for? Exporting and encrypt/decrypt are _not_ mutually-exclusive in HPKE. A single HPKE context is usable for encrypt/decrypt (depending on side) and export.

If you end up having to define multiple contexts, you'll have two problems:
* On the receiver side, you'll end up doing an extra decap operation (the expensive part) for each operation.
* On the sender side, the setup process is non-deterministic, so you won't even get the same context each time anyway! (And, obviously, BoringSSL will not support an API marked _for testing_ being used in production. If we see such abuse, we'll unexport the testing API and keep it private to the library. It is _only_ exported as a convenience for unit tests.)